### ISSUE TITLE
Cannot infer return type on extern.

### ISSUE THREAD

**Issue by hanw** [p4lang/p4c#1973]
The attached program does not compile with p4test.

[sizeof.p4.txt](https://github.com/p4lang/p4c/files/3302929/sizeof.p4.txt)

However, it seems that type inference should be possible on the extern return type in this case?

```
./p4test ../testdata/p4_16_samples/sizeof.p4
../testdata/p4_16_samples/sizeof.p4(25): [--Werror=type-error] error: Cannot infer return type sizeOf
        bit<16> m = sizeOf(h);
                    ^^^^^^^^^
```

**Comment by mbudiu-vmw** [comment]
Indeed, we do not propagate constraints for function return values from the outside.
For such cases you have to specify explicitly the type parameters of the function: `sizeOf<bit<16>, bit<32>>(h);`

**Comment by antoninbas** [comment]
I believe the same goes for C++ templates.

**Comment by hanw** [comment]
Similarly, this does not work either. Are these issues bug to the type inference?

```
extern int sizeOf<T, H>(H h);

header h_t {
    bit<8> f;
}

control ingress(h_t h) {
    apply {
        bit<16> m = (bit<16>)sizeOf(h);
    }
}
``` 

```
sizeof.p4(9): [--Werror=type-error] error: sizeOf: Cannot unify type int with R
        bit<16> m = (bit<16>)sizeOf(h);
                             ^^^^^^^^^
```

**Comment by mbudiu-vmw** [comment]
I would not say that this is a bug, it is a limitation of the type inference by design.
The second error message is not very good though, that could be construed as a bug.

**Comment by mbudiu-vmw** [comment]
By definition `int` means "compile-time constant". So you cannot really have a function that returns an int - unless that function can be evaluated at compile-time.

**Comment by mbudiu-vmw** [comment]
We do not allow run-time values of `int` type, these only exist strictly at compile-time.

**Comment by hanw** [comment]
The `sizeOf` function is intended to be evaluated to a constant at compile-time.

**Comment by mbudiu-vmw** [comment]
This is a tricky one, since the front-end expects to eliminate all `int` values very early, but obviously it does not know anything about this function. We may need a way to inject additional passes in the front-end to teach it about such functions.

**Comment by mbudiu-vmw** [comment]
Alternatively, just make it return a `bit<16>` and do the constant folding in the back-end.

**Comment by jafingerhut** [comment]
Are you actually wanting to implement another function that behaves like the newly added minSizeInBytes(), but with a different name?

Or are you asking more generally about other extern functions that you want to return a compile-time constant value of type `int`?

**Comment by jnfoster** [comment]
☝️

**Comment by hanw** [comment]
I need a `sizeOf` function that is not converted to constant in the frontend as I have mentioned in the `minSizeInBytes` PR. The way `minSizeInBytes` is implemented today, it does not work for our use case.

**Comment by hanw** [comment]
I would also need a way to generate an error if people use `minSizeInBytes` on certain headers in the frontend. That is a separate issue.

**Comment by hesingh** [comment]
There is `minWidthBits` in frontend code (`p4c/frontends/p4/typeMap.cpp`).  You could change the implementation to return 0 for "not supported" or "-1" for failure for your asic.

**Comment by cc10512** [comment]
We should consider adding support for compiler intrinsics -- which can be specialized per target, rather than forcing certain behaviors directly in the frontend. This will enable backends to understand the semantics of these calls and do different optimizations that are simply not possible in the frontend.

**Comment by jafingerhut** [comment]
You may have reasons I am unaware of to make the return type a type parameter, but it seems that you could hard-code it to something like `bit<16>`, or whatever type is most natural for your target, and let users cast it down to smaller sizes if they want.

**Comment by cc10512** [comment]
It is not the type (or it's width) but rather the value. If the frontend replaces it with a constant, the backend no longer knows that this was intended to be the function that counts the size of a struct. And if the backend wants to change the layout/padding of the struct, the value is incorrect.

**Comment by hanw** [comment]
@jafingerhut has a good suggestion. We will need a separate extern to deal with the `special` header we have, but we can live with that.

@hesingh I don't think I could override the implementation of `minWidthbits` in its current form.

**Comment by jafingerhut** [comment]
@cc10512 It is only a return type of `int` that is forced to be a compile-time constant value.

If the return type is `bit<16>` (whether that is a type parameter or not), the compiler should have no reason to assume the return value is a compile-time constant, unless you do special fiddling about in the compiler to make it believe it is a compile time constant somehow.
